wallets/validate.js

```diff
@@ -0,0 +1,110 @@
+/*
+   we want to take all the validate members from the provided wallet
+   and compose into a single yup schema for formik validation ...
+   the validate member can be on of:
+   - a yup schema
+   - a function that throws on an invalid value
+   - a regular expression that must match
+*/
+
+import { autowithdrawSchemaMembers, vaultEntrySchema } from '@/lib/validate'
+import * as Yup from '@/lib/yup'
+import { canReceive } from './common'
+
+export default async function validateWallet (walletDef, data,
+  { yupOptions = { abortEarly: true }, topLevel = true, serverSide = false, skipGenerated = false } = {}) {
+  let schema = composeWalletSchema(walletDef, serverSide, skipGenerated)
+
+  if (canReceive({ def: walletDef, config: data })) {
+    schema = schema.concat(autowithdrawSchemaMembers)
+  }
+
+  await schema.validate(data, yupOptions)
+
+  const casted = schema.cast(data, { assert: false })
+  if (topLevel && walletDef.validate) {
+    await walletDef.validate(casted)
+  }
+
+  return casted
+}
+
+function createFieldSchema (name, validate) {
+  if (!validate) {
+    throw new Error(`No validation provided for field ${name}`)
+  }
+
+  if (Yup.isSchema(validate)) {
+    // If validate is already a Yup schema, return it directly
+    return validate
+  } else if (typeof validate === 'function') {
+    // If validate is a function, create a custom Yup test
+    return Yup.mixed().test({
+      name,
+      test: (value, context) => {
+        try {
+          validate(value)
+          return true
+        } catch (error) {
+          return context.createError({ message: error.message })
+        }
+      }
+    })
+  } else if (validate instanceof RegExp) {
+    // If validate is a regular expression, use Yup.matches
+    return Yup.string().matches(validate, `${name} is invalid`)
+  } else {
+    throw new Error(`validate for ${name} must be a yup schema, function, or regular expression`)
+  }
+}
+
+function composeWalletSchema (walletDef, serverSide, skipGenerated) {
+  const { fields } = walletDef
+
+  const vaultEntrySchemas = { required: [], optional: [] }
+  const schemaShape = fields.reduce((acc, field) => {
+    const { name, validate, optional, generated, clientOnly, requiredWithout } = field
+
+    if (generated && skipGenerated) {
```

Comment:
generated fields (LNC dh keys) are selectively skipped until we know they've been generated (ie after testing a payment/connection).